File: Algorithm/BinaryTreePostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// // Solution 1: 
// class Solution {
//     public List<Integer> postorderTraversal(TreeNode root) {
//         List<Integer> ans = new ArrayList<Integer>();
//         postorder(root, ans);
//         return ans;
//     }
//     private void postorder(TreeNode root, List<Integer> ans) {
//         if (root == null) return;
//         postorder(root.left, ans);
//         postorder(root.right, ans);
//         ans.add(root.val);
//     }
// }

// // Solution 2:
// class Solution {
//     public List<Integer> postorderTraversal(TreeNode root) {
//         if (root == null) return new ArrayList<Integer>();
//         List<Integer> ans = new ArrayList<Integer>();
//         List<Integer> l = postorderTraversal(root.left);
//         List<Integer> r = postorderTraversal(root.right);
//         ans.addAll(l);
//         ans.addAll(r);
//         ans.add(root.val);
//         return ans;
//     }
// }

// Solution 3:
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        if (root == null) return new ArrayList<>();
        Deque<Integer> ans = new LinkedList<>();
        Stack<TreeNode> s = new Stack<>();
        s.push(root);
        while (!s.isEmpty()) {
            TreeNode n = s.pop();
            ans.addFirst(n.val); // 0(1)
            if (n.left != null) s.push(n.left);
            if (n.right != null) s.push(n.right);
        }
        return new ArrayList<Integer>(ans);
    }
}

// class Solution {
//     public List<Integer> postorderTraversal(TreeNode root) {
//         if (root == null) return new ArrayList<Integer>();
//         List<Integer> ans = new ArrayList<>();
//         Stack<TreeNode> s = new Stack<>();
//         s.push(root);
//         while (!s.isEmpty()) {
//             TreeNode n = s.pop();
//             ans.add(n.val); // 这里时间为O(n) C++ 的queue方法 push_front为 0(1)。
//             if (n.left != null) s.push(n.left);
//             if (n.right != null) s.push(n.right);
//         }
//         Collections.reverse(ans);
//         return ans;
//     }
// }

File: Algorithm/BinaryTreePostorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// // Solution 1:
// class Solution {
// public:
//     vector<int> postorderTraversal(TreeNode* root) {
//         vector<int> ans;
//         postorderTraversal(root, ans);
//         return ans;
//     }

//     void postorderTraversal(TreeNode* root, vector<int>& ans) {
//         if (!root) return;
//         postorderTraversal(root->left, ans);
//         postorderTraversal(root->right, ans);
//         ans.push_back(root->val);
//     }
// };

// // Solution 2:
// class Solution {
// public:
//     vector<int> postorderTraversal(TreeNode* root) {
//         if (!root) return {};
//         vector<int> ans;
//         const vector<int> l = postorderTraversal(root->left);
//         const vector<int> r = postorderTraversal(root->right);
//         ans.insert(ans.end(), l.begin(), l.end());
//         ans.insert(ans.end(), r.begin(), r.end());
//         ans.push_back(root->val);
//         return ans;
//     }
// };

// Solution 3:
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if (!root) return {};
        deque<int> ans;
        stack<TreeNode*> s;
        s.push(root);
        while (!s.empty()) {
            TreeNode* n = s.top();
            s.pop();
            ans.push_front(n->val); // O(1)
            if (n->left) s.push(n->left);
            if (n->right) s.push(n->right);
        }
        return vector<int>(ans.begin(), ans.end());
    }
};
